crates/router/src/services/authentication.rs

```diff
@@ -1033,6 +1033,78 @@ where
     }
 }
 
+pub struct JWTAuthProfileFromRoute {
+    pub profile_id: Option<String>,
+    pub required_permission: Permission,
+}
+
+#[async_trait]
+impl<A> AuthenticateAndFetch<AuthenticationData, A> for JWTAuthProfileFromRoute
+where
+    A: SessionStateInfo + Sync,
+{
+    async fn authenticate_and_fetch(
+        &self,
+        request_headers: &HeaderMap,
+        state: &A,
+    ) -> RouterResult<(AuthenticationData, AuthenticationType)> {
+        let payload = parse_jwt_payload::<A, AuthToken>(request_headers, state).await?;
+        if payload.check_in_blacklist(state).await? {
+            return Err(errors::ApiErrorResponse::InvalidJwtToken.into());
+        }
+
+        let permissions = authorization::get_permissions(state, &payload).await?;
+        authorization::check_authorization(&self.required_permission, &permissions)?;
+        let key_manager_state = &(&state.session_state()).into();
+        let key_store = state
+            .store()
+            .get_merchant_key_store_by_merchant_id(
+                key_manager_state,
+                &payload.merchant_id,
+                &state.store().get_master_key().to_vec().into(),
+            )
+            .await
+            .to_not_found_response(errors::ApiErrorResponse::InvalidJwtToken)
+            .attach_printable("Failed to fetch merchant key store for the merchant id")?;
+
+        if let Some(ref payload_profile_id) = payload.profile_id {
+            if let Some(ref profile_id) = self.profile_id {
+                if *payload_profile_id != *profile_id {
+                    return Err(report!(errors::ApiErrorResponse::InvalidJwtToken));
+                } else {
+                    let merchant = state
+                        .store()
+                        .find_merchant_account_by_merchant_id(
+                            key_manager_state,
+                            &payload.merchant_id,
+                            &key_store,
+                        )
+                        .await
+                        .to_not_found_response(errors::ApiErrorResponse::InvalidJwtToken)
+                        .attach_printable("Failed to fetch merchant account for the merchant id")?;
+
+                    let auth = AuthenticationData {
+                        merchant_account: merchant,
+                        key_store,
+                        profile_id: payload.profile_id,
+                    };
+                    Ok((
+                        auth.clone(),
+                        AuthenticationType::MerchantJwt {
+                            merchant_id: auth.merchant_account.get_id().clone(),
+                            user_id: Some(payload.user_id),
+                        },
+                    ))
+                }
+            } else {
+                return Err(report!(errors::ApiErrorResponse::InvalidJwtToken));
+            }
+        } else {
+            return Err(report!(errors::ApiErrorResponse::InvalidJwtToken));
+        }
```

Comment:
as per this logic, authentication will fail if either of the 2 profile_ids is None.

Reply:
This is fixed as now, we are handling it in different way.